File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> userInfo = new HashMap<>();
        List<String[]> actions = new ArrayList<>();
        
        for(String str : record){
            String[] strArray = str.split(" ");
            
            String cmd = strArray[0];
            String uid = strArray[1];
            
            if(cmd.equals("Enter") || cmd.equals("Change")) userInfo.put(uid, strArray[2]);
            if(cmd.equals("Enter") || cmd.equals("Leave")) actions.add(new String[]{cmd, uid});
        }
        
        List<String> result = new ArrayList<>();
        for(String[] action : actions){
            String cmd = action[0];
            String uid = action[1];
            
            if(cmd.equals("Enter")) result.add(userInfo.get(uid) + "님이 들어왔습니다.");
            else result.add(userInfo.get(uid) + "님이 나갔습니다.");
        }
        
        return result.toArray(String[]::new);
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    user_info = {}
    action = []
    
    for r in record:
        info = r.split()
        cmd, user_id = info[0], info[1]
        
        if cmd in ("Enter", "Change"):
            user_info[user_id] = info[2]
            
        if cmd in ("Enter", "Leave"):
            action.append((cmd, user_id))
    
    result = []
    for cmd, user_id in action:
        if cmd == "Enter":
            result.append(f"{user_info[user_id]}님이 들어왔습니다.")
        else:
            result.append(f"{user_info[user_id]}님이 나갔습니다.")
                          
    return result